File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/LastMatchedTradedVolumeCategorizer.java
package cz.fb.manaus.reactor.categorizer;

import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.category.categorizer.SettledBetCategorizer;
import cz.fb.manaus.core.model.BetAction;
import cz.fb.manaus.core.model.MarketPrices;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.SettledBet;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.OptionalDouble;
import java.util.Set;

@Component
public class LastMatchedTradedVolumeCategorizer implements SettledBetCategorizer {

    public static final String LAST_MATCHED = "lastMatched";


    private String getCategory(double tradedMean, double lastMatched) {
        if (Price.priceEq(tradedMean, lastMatched)) {
            return LAST_MATCHED + "_eqTradedMean";
        } else if (lastMatched > tradedMean) {
            return LAST_MATCHED + "_gtTradedMean";
        } else {
            return LAST_MATCHED + "_ltTradedMean";
        }
    }

    @Override
    public boolean isMarketSnapshotRequired() {
        return true;
    }

    @Override
    public Set<String> getCategories(SettledBet settledBet, BetCoverage coverage) {
        BetAction action = settledBet.getBetAction();
        OptionalDouble tradedMean = action.getDoubleProperty(BetAction.TRADED_VOL_MEAN);
        if (tradedMean.isPresent()) {
            MarketPrices marketPrices = settledBet.getBetAction().getMarketPrices();
            double lastMatchedPrice = marketPrices.getRunnerPrices(settledBet.getSelectionId()).getLastMatchedPrice();
            return Collections.singleton(getCategory(tradedMean.getAsDouble(), lastMatchedPrice));
        } else {
            return Set.of();
        }
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/LastMatchedTradedVolumeCategorizer.kt
package cz.fb.manaus.reactor.categorizer

import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.category.categorizer.RealizedBetCategorizer
import cz.fb.manaus.core.repository.domain.BetAction
import cz.fb.manaus.core.repository.domain.Price
import cz.fb.manaus.core.repository.domain.RealizedBet
import org.springframework.stereotype.Component

@Component
class LastMatchedTradedVolumeCategorizer : RealizedBetCategorizer {

    private fun getCategory(tradedMean: Double, lastMatched: Double): String {
        return when {
            Price.priceEq(tradedMean, lastMatched) -> LAST_MATCHED + "_eqTradedMean"
            lastMatched > tradedMean -> LAST_MATCHED + "_gtTradedMean"
            else -> LAST_MATCHED + "_ltTradedMean"
        }
    }

    override val isMarketSnapshotRequired: Boolean = true

    override fun getCategories(realizedBet: RealizedBet, coverage: BetCoverage): Set<String> {
        val action = realizedBet.betAction
        val tradedMean = action.getDoubleProperty(BetAction.TRADED_VOL_MEAN)
        return if (tradedMean.isPresent) {
            val marketPrices = realizedBet.betAction.marketPrices
            val lastMatchedPrice = marketPrices.getRunnerPrices(realizedBet.selectionId).lastMatchedPrice
            setOf(getCategory(tradedMean.asDouble, lastMatchedPrice!!))
        } else {
            emptySet()
        }
    }

    companion object {
        private const val LAST_MATCHED = "lastMatched"
    }
}
